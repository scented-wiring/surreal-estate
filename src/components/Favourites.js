import React, { useState } from "react";
import "../styles/Favourites.css";
import PropTypes from "prop-types";
import Alert from "./Alert";
import axios from "axios";
import FavouriteCard from "./FavouriteCard";
import { useEffect } from "react";

const Favourites = ({ userID }) => {
  const [favourites, setFavourites] = useState([]);
  const [alert, setAlert] = useState({ message: "", isSuccess: false });

  useEffect(() => {
    let isMounted = true;
    axios
      .get("http://localhost:4000/api/v1/Favourite")
      .then((response) => {
        if (isMounted) setFavourites(response.data);
      })
      .catch(() => {
        setAlert({
          message: "Could not connect to server.",
          isSuccess: false,
        });
      });
    return () => {
      isMounted = false;
    };
  }, []);

  const handleRemoveFavourite = (_id) => {
    axios
      .delete(`http://localhost:4000/api/v1/Favourite/${_id}`)
      .then(() => {
        setAlert({
          message: "Favourite removed",
          isSuccess: true,
        });
      })
      .then(() => setFavourites(favourites.filter((fav) => fav._id !== _id)))
      .then(() => {
        setTimeout(() => setAlert({ message: "", isSuccess: false }), 2000);
      })
      .catch(() => {
        setAlert({
          message: "Server error",
          isSuccess: false,
        });
      });
  };

  if (alert.message === "Could not connect to server.") {
    return (
      <div className="favourites">
        <Alert message={alert.message} success={alert.isSuccess} />
      </div>
    );
  } else if (!userID || favourites.length === 0) {
    return (
      <div className="favourites">
        <Alert message={alert.message} success={alert.isSuccess} />
        <div className="no-favourites">
          No favourites found. Log in with Facebook to select and view
          favourites.
        </div>
      </div>
    );
  } else {
    return (
      <div className="favourites">
        <Alert message={alert.message} success={alert.isSuccess} />
        {favourites.map((favourite) => (
          <div key={favourite._id}>
            <FavouriteCard
              propertyListing={favourite.propertyListing}
              _id={favourite._id}
              removeFavourite={handleRemoveFavourite}
            />
          </div>
        ))}
      </div>
    );
  }
};

Favourites.propTypes = {
  userID: PropTypes.string.isRequired,
};

export default Favourites;
