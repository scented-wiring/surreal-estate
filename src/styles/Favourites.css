.favourites {
  display: flex;
  flex: 12;
  flex-direction: column;
  align-items: center;
  padding: 0 20px 0 20px;
}

.no-favourites {
  margin-top: 10px;
}

@media only screen and (max-width: 469px) {
  .no-favourites {
    font-size: 0.9em;
  }
}
