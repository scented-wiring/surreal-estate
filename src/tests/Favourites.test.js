import React from "react";
import { render } from "@testing-library/react";
import Favourites from "../components/Favourites";

describe("Favourites", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<Favourites />);
    expect(asFragment).toMatchSnapshot();
  });

  it("displays a message if userID is null", () => {
    const { getByText } = render(<Favourites />);
    expect(getByText(/No favourites found/)).toBeInTheDocument();
  });
});
