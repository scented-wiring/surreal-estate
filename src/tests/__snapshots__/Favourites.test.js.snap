// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Favourites renders a component 1`] = `[Function]`;

exports[`Favourites renders correctly 1`] = `[Function]`;
